providers/common/provider_util.c

```diff
@@ -0,0 +1,141 @@
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/evp.h>
+#include "internal/provider_util.h"
+
+void ossl_prov_cipher_free(PROV_CIPHER *pc)
+{
+    EVP_CIPHER_free(pc->alloc_cipher);
+    pc->alloc_cipher = NULL;
+    pc->cipher = NULL;
+    pc->engine = NULL;
+}
+
+int ossl_prov_cipher_dup(PROV_CIPHER *dst, const PROV_CIPHER *src)
+{
+    if (src->alloc_cipher != NULL && !EVP_CIPHER_up_ref(src->alloc_cipher))
+        return 0;
+    dst->engine = src->engine;
+    dst->cipher = src->cipher;
+    dst->alloc_cipher = src->alloc_cipher;
+    return 1;
+}
+
+static const OSSL_PARAM *load_common(const OSSL_PARAM params[],
+                                     const char **propquery, ENGINE **engine,
+                                     const char *pname)
+{
+    const OSSL_PARAM *p;
+
+    *propquery = NULL;
+    p = OSSL_PARAM_locate_const(params, "properties");
+    if (p != NULL) {
+        if (p->data_type != OSSL_PARAM_UTF8_STRING)
+            return NULL;
+        *propquery = p->data;
+    }
+
+    *engine = NULL;
+    /* TODO legacy stuff, to be removed */
+#ifndef FIPS_MODE /* Inside the FIPS module, we don't support legacy ciphers */
+    p = OSSL_PARAM_locate_const(params, "engine");
+    if (p != NULL) {
+        if (p->data_type != OSSL_PARAM_UTF8_STRING)
+            return NULL;
+        ENGINE_finish(*engine);
+        *engine = ENGINE_by_id(p->data);
+        if (*engine == NULL)
+            return NULL;
+    }
+#endif
+    p = OSSL_PARAM_locate_const(params, pname);
+    if (p->data_type != OSSL_PARAM_UTF8_STRING)
+        return NULL;
+    return p;
+}
+
+int ossl_prov_cipher_load_from_params(PROV_CIPHER *pc,
+                                      const OSSL_PARAM params[],
+                                      OPENSSL_CTX *ctx)
+{
+    const OSSL_PARAM *p;
+    const char *propquery;
+
+    if ((p = load_common(params, &propquery, &pc->engine, "cipher")) == NULL)
+        return 0;
+
+    EVP_CIPHER_free(pc->alloc_cipher);
+    pc->cipher = pc->alloc_cipher = EVP_CIPHER_fetch(ctx, p->data, propquery);
+    /* TODO legacy stuff, to be removed */
+#ifndef FIPS_MODE /* Inside the FIPS module, we don't support legacy ciphers */
+    if (pc->cipher == NULL)
+        pc->cipher = EVP_get_cipherbyname(p->data);
+#endif
+    return pc->cipher != NULL;
+}
+
+const EVP_CIPHER *ossl_prov_cipher_cipher(const PROV_CIPHER *pc)
+{
+    return pc->cipher;
+}
+
+ENGINE *ossl_prov_cipher_engine(const PROV_CIPHER *pc)
+{
+    return pc->engine;
+}
+
+void ossl_prov_digest_free(PROV_DIGEST *pd)
+{
+    EVP_MD_free(pd->alloc_md);
+    pd->alloc_md = NULL;
+    pd->md = NULL;
+    pd->engine = NULL;
+}
+
+int ossl_prov_digest_dup(PROV_DIGEST *dst, const PROV_DIGEST *src)
```

Comment:
_dup is again not right as this does not provide the allocation - should be _copy instead.

Reply:
Good idea.

Reply:
What about _clone ?

It isn't quite a straight copy.

Reply:
We already have `_copy` as a pattern, and they aren't straight copy either, there are some fields that get re-allocate as needed.

Reply:
Changed to _copy.